### PR TITLE
feat(core): docanchor and graph architypes

### PR DESCRIPTION
**Pull request by ypkang**

### CODE REVIEW

File: jaclang_jaseci/plugins/common.py [hunk 1]
```diff
@@ -0,0 +1,1174 @@
...
+    NodeAnchor as _NodeAnchor,
+    NodeArchitype as _NodeArchitype,
+    Root as _Root,
+)
+
+from motor.motor_asyncio import AsyncIOMotorClientSession
+
+from orjson import dumps
+
+from pymongo import ASCENDING, DeleteMany, DeleteOne, InsertOne, UpdateMany, UpdateOne
```
Comment (ypkang) [inline jaclang_jaseci/plugins/common.py:39]: why does this need to access mongodb directly, as opposed to through collections?

File: jaclang_jaseci/plugins/common.py [hunk 2]
```diff
@@ -0,0 +1,1174 @@
...
+
+
+class ArchCollection(BaseCollection[T]):
+    """Default Collection for Architypes."""
+
+    @classmethod
+    def build_node(
+        cls, doc_anc: "DocAnchor[NodeArchitype]", doc: Mapping[str, Any]
+    ) -> "NodeArchitype":
+        """Translate EdgeArchitypes edges into DocAnchor edges."""
```
Comment (ypkang) [inline jaclang_jaseci/plugins/common.py:65]: This docstring is not accurate.

File: jaclang_jaseci/plugins/common.py [hunk 3]
```diff
@@ -0,0 +1,1174 @@
...
+            arch._jac_.target = DocAnchor[NodeArchitype].ref(tgt)
+
+        arch._jac_.is_undirected = bool(doc.get("is_undirected"))
+
+        return arch
+
+
+@dataclass
+class DocAccess:
+    """DocAnchor for Access Handler."""
```
Comment (ypkang) [inline jaclang_jaseci/plugins/common.py:94]: Let's add some description here on how the access json is structured and how it works. This will make it easier for debugging later.

File: jaclang_jaseci/plugins/common.py [hunk 4]
```diff
@@ -0,0 +1,1174 @@
...
+    id: ObjectId = field(default_factory=ObjectId)
+    root: Optional[ObjectId] = None
+    access: DocAccess = field(default_factory=DocAccess)
+    connected: bool = False
+    arch: Optional["DocArchitype[DA]"] = None
+    changes: dict[str, dict[str, Any]] = field(default_factory=dict)
+    hashes: dict[str, int] = field(default_factory=dict)
+    rollback_changes: dict[str, dict[str, Any]] = field(default_factory=dict)
+    rollback_hashes: dict[str, int] = field(default_factory=dict)
+    current_access_level: Optional[int] = None
```
Comment (ypkang) [inline jaclang_jaseci/plugins/common.py:141]: Let's add a description for each of these attributes, for readability.

File: jaclang_jaseci/plugins/common.py [hunk 5]
```diff
@@ -0,0 +1,1174 @@
...
+            pull[field] = {"$in": set()}
+
+        ops: set = pull[field]["$in"]
+
+        if remove:
+            if obj in ops:
+                ops.remove(obj)
+        else:
+            ops.add(obj)
+            self._add_to_set(field, obj, True)
```
Comment (ypkang) [inline jaclang_jaseci/plugins/common.py:188]: i am a bit confused about _add_to_set and _pull here. Why would they call each other?

File: jaclang_jaseci/plugins/common.py [hunk 6]
```diff
@@ -0,0 +1,1174 @@
...
+
+    def connect_edge(self, doc_anc: "DocAnchor[DA]", rollback: bool = False) -> None:
+        """Push update that there's newly added edge."""
+        if not rollback:
+            self._add_to_set("edge", doc_anc)
+        else:
+            self._pull("edge", doc_anc, True)
+
+    def disconnect_edge(self, doc_anc: "DocAnchor[DA]") -> None:
```
Comment (ypkang) [inline jaclang_jaseci/plugins/common.py:197]: why doesn't disconnect_edge need rollback?

File: jaclang_jaseci/plugins/common.py [hunk 7]
```diff
@@ -0,0 +1,1174 @@
...
+        """Push update that there's edge that has been removed."""
+        self._pull("edge", doc_anc)
+
+    def allow_node(self, node_id: ObjectId, write: bool = False) -> None:
+        """Allow target node to access current Architype."""
+        w = 1 if write else 0
```
Comment (ypkang) [inline jaclang_jaseci/plugins/common.py:203]: I'd use enum for the access indicators. more readable.

File: jaclang_jaseci/plugins/common.py [hunk 8]
```diff
@@ -0,0 +1,1174 @@
...
+        return {
+            "_id": self.id,
+            "name": self.name,
+            "root": self.root,
+            "access": self.access.json(),
+        }
+
+    @classmethod
+    def ref(cls, ref_id: str) -> Optional["DocAnchor[DA]"]:
+        """Return DocAnchor instance if ."""
```
Comment (ypkang) [inline jaclang_jaseci/plugins/common.py:293]: Complete this docstring.

File: jaclang_jaseci/plugins/common.py [hunk 9]
```diff
@@ -0,0 +1,1174 @@
...
+
+class DocArchitype(Generic[DA]):
+    """DocAnchor Class Handler."""
+
+    _jac_type_: JType
+
+    class Collection(ArchCollection):
+        """Default DocArchitype Collection."""
+
+        pass
```
Comment (ypkang) [inline jaclang_jaseci/plugins/common.py:355]: Is this required? since it is not actually overriding anything.

File: jaclang_jaseci/plugins/common.py [hunk 10]
```diff
@@ -0,0 +1,1174 @@
...
+        ctx = {}
+        if is_dataclass(self) and not isinstance(self, type):
+            ctx = asdict(self)
+
+        return {"context": ctx}
+
+    async def propagate_save(
+        self, changes: dict, session: AsyncIOMotorClientSession
+    ) -> list[InsertOne[Any] | DeleteMany | DeleteOne | UpdateMany | UpdateOne]:
+        """Propagate saving."""
```
Comment (ypkang) [inline jaclang_jaseci/plugins/common.py:388]: need help understanding this function.

File: jaclang_jaseci/plugins/common.py [hunk 11]
```diff
@@ -0,0 +1,1174 @@
...
+    async def save_with_session(self) -> None:
+        """Upsert Architype with session."""
+        async with await ArchCollection.get_session() as session:
+            async with session.start_transaction():
+                try:
+                    await self.save(session)
+                    await session.commit_transaction()
+                except Exception:
+                    await session.abort_transaction()
+                    logger.exception("Error saving node!")
```
Comment (ypkang) [inline jaclang_jaseci/plugins/common.py:433]: This is not accurate right? This won't neccessarily be just node.

File: jaclang_jaseci/plugins/common.py [hunk 12]
```diff
@@ -0,0 +1,1174 @@
...
+        """Upsert Architype."""
+        raise Exception("Save function not implemented yet!")
+
+    async def save_with_session(self) -> None:
+        """Upsert Architype with session."""
+        async with await ArchCollection.get_session() as session:
+            async with session.start_transaction():
+                try:
+                    await self.save(session)
+                    await session.commit_transaction()
```
Comment (ypkang) [inline jaclang_jaseci/plugins/common.py:430]: If i understand this correctly, all nodes will be saved in one commit and all edges will be saved in a separate commit. is that correct?

Comment (ypkang) [reply jaclang_jaseci/plugins/common.py:430]: And any delete will be in another commit, handled by `destroy`?

File: jaclang_jaseci/plugins/common.py [hunk 13]
```diff
@@ -0,0 +1,1174 @@
...
+            not jctx
+            or not (from_jd := self._jac_doc_).connected
+            or not (to_jd := to._jac_doc_).connected
+            or not (from_root := from_jd.root)
+            or not (to_root := to_jd.root)
+        ):
+            return False
+
+        if (isinstance(self, Root) and from_jd.id == to_root) or from_root == to_root:
+            to_jd.current_access_level = 1
```
Comment (ypkang) [inline jaclang_jaseci/plugins/common.py:468]: I'd make the access level an enum as well. More readable.

File: jaclang_jaseci/plugins/common.py [hunk 14]
```diff
@@ -0,0 +1,1174 @@
...
+                return True
+
+        if isinstance(
+            to_root_access := await jctx.check_root_access(to_root),
+            DocAccess,
+        ) and (cur_root_id := jctx.get_root_id()):
+            for i in range(1, -1, -1):
+                if cur_root_id in to_root_access.roots[i]:
+                    to_jd.current_access_level = i
+                    return True
```
Comment (ypkang) [inline jaclang_jaseci/plugins/common.py:487]: need help understanding this piece of logic.

File: jaclang_jaseci/plugins/common.py [hunk 15]
```diff
@@ -0,0 +1,1174 @@
...
+class NodeArchitype(_NodeArchitype, DocArchitype["NodeArchitype"]):
+    """Overriden NodeArchitype."""
+
+    _jac_type_ = JType.node
+
+    def __init__(self) -> None:
+        """Create node architype."""
+        self._jac_: NodeAnchor = NodeAnchor(obj=self)
+
+    async def connect(self) -> "NodeArchitype":
```
Comment (ypkang) [inline jaclang_jaseci/plugins/common.py:515]: is this needed?

File: jaclang_jaseci/plugins/common.py [hunk 16]
```diff
@@ -0,0 +1,1174 @@
...
+        """Return self."""
+        return self
+
+    class Collection(ArchCollection["NodeArchitype"]):
+        """Default NodeArchitype Collection."""
+
+        __collection__ = "node"
+        __indexes__ = [
+            {"fields": [("_id", ASCENDING), ("name", ASCENDING), ("root", ASCENDING)]}
+        ]
```
Comment (ypkang) [inline jaclang_jaseci/plugins/common.py:525]: We should document the key-value structure for each architype (node, edge, user) somewhere. Maybe not in the code here but in a documentation markdown somewhere, as part of this PR.

File: jaclang_jaseci/plugins/common.py [hunk 17]
```diff
@@ -0,0 +1,1174 @@
...
+        self._jac_doc_.disconnect_edge(edge._jac_doc_)
+
+    def allow_node(
+        self, node: Union[DocAnchor, "NodeArchitype"], write: bool = False
+    ) -> None:
+        """Allow target node to access current Architype."""
+        if isinstance(node, DocAnchor):
+            self._jac_doc_.allow_node(node.id, write)
+        elif isinstance(node, NodeArchitype):
+            self._jac_doc_.allow_node(node._jac_doc_.id, write)
```
Comment (ypkang) [inline jaclang_jaseci/plugins/common.py:561]: hmmm i think we need to have a diagram for these concepts and their relationships, docanchor, nodearchitype, docarchitype, nodeanchor etc....

File: jaclang_jaseci/plugins/common.py [hunk 18]
```diff
@@ -0,0 +1,1174 @@
...
+
+        # allowed to throw error to stop executions
+        async for edge in (await NodeArchitype.Collection.collection()).find(
+            {"_id": {"$in": list(queue.keys())}}
+        ):
+            if cls := queue.get(edge["_id"]):
+                arch = cls.Collection.__document__(edge)
+                self.set(arch._jac_doc_.id, arch)
+
+    async def old_populate(
```
Comment (ypkang) [inline jaclang_jaseci/plugins/common.py:1069]: still needed?

File: jaclang_jaseci/plugins/common.py [hunk 19]
```diff
@@ -0,0 +1,1174 @@
...
+base_root = Root.Collection.build_node(
+    DocAnchor[Root](
+        type=JType.node,
+        id=base_root_id,
+        root=base_root_id,
+        access=DocAccess(),
+        connected=True,
+    ),
+    {"context": {}, "edge": []},
+)
```
Comment (ypkang) [inline jaclang_jaseci/plugins/common.py:1174]: Are these (after line 1159) hardcoded to create root node? shouldn't be done per user and upon user creation?

File: jaclang_jaseci/plugins/common.py (hunk 1 above)
Comment (amadolid) [reply jaclang_jaseci/plugins/common.py:39]: These are just operation models and not yet been executed. It will be used for bulk writing thru collections.

File: jaclang_jaseci/plugins/common.py (hunk 5 above)
Comment (amadolid) [reply jaclang_jaseci/plugins/common.py:188]: this is to align addToSet and pull request.

the parameter remove will control it
if _add_to_set is triggered the _pull request should not include that object too
vice versa

the object should be on either addToSet or pull never both

File: jaclang_jaseci/plugins/common.py (hunk 15 above)
Comment (amadolid) [reply jaclang_jaseci/plugins/common.py:515]: This is to fix hinting and linting for mypy/flake8

File: jaclang_jaseci/plugins/common.py (hunk 16 above)
Comment (amadolid) [reply jaclang_jaseci/plugins/common.py:525]: will add it

File: jaclang_jaseci/plugins/common.py (hunk 18 above)
Comment (amadolid) [reply jaclang_jaseci/plugins/common.py:1069]: will remove

File: jaclang_jaseci/plugins/common.py (hunk 19 above)
Comment (amadolid) [reply jaclang_jaseci/plugins/common.py:1174]: This is for Generic Root instance used for non authenticated users
